gc.c

```diff
@@ -3451,13 +3453,28 @@ Init_heap(void)
 #endif
 
     heap_add_pages(objspace, &size_pools[0], SIZE_POOL_EDEN_HEAP(&size_pools[0]), gc_params.heap_init_slots / HEAP_PAGE_OBJ_LIMIT);
+    size_pools[0].next = NULL;
 
-    /* Give other size pools allocatable pages. */
-    for (int i = 1; i < SIZE_POOL_COUNT; i++) {
-        rb_size_pool_t *size_pool = &size_pools[i];
-        int multiple = size_pool->slot_size / sizeof(RVALUE);
-        size_pool->allocatable_pages = gc_params.heap_init_slots * multiple / HEAP_PAGE_OBJ_LIMIT;
+    if (SIZE_POOL_COUNT > 1) {
+        size_pools[0].next = &size_pools[1];
+
+        for (int i = 1; i < SIZE_POOL_COUNT; i++) {
+            rb_size_pool_t *size_pool = &size_pools[i];
+            int multiple = size_pool->slot_size / sizeof(RVALUE);
+            size_pool->allocatable_pages = gc_params.heap_init_slots * multiple / HEAP_PAGE_OBJ_LIMIT;
+            /* set up the links between pools */
+            if (i == (SIZE_POOL_COUNT - 1)) {
+                /* make sure the last pool points at nothing */
+                size_pool->next = NULL;
+            }
+            else {
+                GC_ASSERT(size_pool != &size_pools[++i]);
+                size_pool->next = &size_pools[++i];
```

Comment:
Why is this incrementing `i`? Wouldn't this mean some size pools are skipped?

Reply:
Whoops. Fixed. Thanks